File: app/src/main/java/com/masum/pdfdownloader/MainActivity.java
package com.masum.pdfdownloader;

import android.app.Dialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.Log;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.masum.pdfdownloader.databinding.ActivityMainBinding;
import com.masum.pdfdownloader.databinding.DownloadDialogBinding;
import com.masum.pdfdownloader.others.Utils;
import com.masum.pdfdownloader.services.ForgroundService;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;

public class MainActivity extends AppCompatActivity implements  ItemClick {
    ActivityMainBinding binding;
    ArrayList<Item>list;
    Boolean isRunning=false;
    DownloadDialogBinding layoutBinding;
    Dialog dialog;
    BroadcastReceiver mMessageReceiver;
    Double fileSize=0.0;
    Boolean isComplete=true;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding=ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        layoutBinding=DownloadDialogBinding.inflate(getLayoutInflater());
        showDialog();

        list= new ArrayList<>();
        list.add(new Item("Test Name", "https://jsoncompare.org/LearningContainer/SampleFiles/PDF/sample-pdf-with-images.pdf"));
        list.add(new Item("File Sample 150kb","https://file-examples-com.github.io/uploads/2017/10/file-sample_150kB.pdf"));
        list.add(new Item("Rahe Belayet","https://www.pixelapps.info/my_important_files/pdf_files/rahe_belayat.pdf"));

        binding.recyclerview.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerview.setAdapter(new RecyclerItemAdapter(list,this));

        mMessageReceiver=new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {

                Double current = intent.getDoubleExtra("timer", 0.0);
                int progress= (int) ((current/(fileSize))*100);

                if (dialog!=null&&layoutBinding!=null){
                    layoutBinding.progressBar.setIndeterminate(false);
                    layoutBinding.progressText.setText(current+" Mb/"+fileSize+"Mb");
                    layoutBinding.progressBar.setProgress(progress);
                }

                Boolean error=intent.getBooleanExtra("error",false);
                if (error&&dialog!=null){
                    dialog.dismiss();
                    Toast.makeText(MainActivity.this,"Fail to download",Toast.LENGTH_SHORT).show();
                }
                Boolean complete=intent.getBooleanExtra("complete",false);
                if (complete){
                    if (binding.recyclerview.getAdapter()!=null)  binding.recyclerview.getAdapter().notifyDataSetChanged();
                    isComplete=true;


                    File file=new File(ForgroundService.path);
                    if (file.exists())
                    {
                        if (isRunning&&dialog.isShowing()){
                            if (dialog!=null &&dialog.isShowing())
                                dialog.dismiss();

                            startActivity(
                                    new Intent(MainActivity.this, PdfActivity.class).putExtra(
                                    "path",
                                    file.getAbsolutePath()
                                    )

                           );
                        }
                        else Toast.makeText(MainActivity.this,"Download Complete",Toast.LENGTH_SHORT).show();
                    }
                }
                else{
                    isComplete=false;
                }



            }
        };

    }

    private void showDialog() {
        dialog = new Dialog(this);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);

        dialog.setContentView(layoutBinding.getRoot());
        dialog.setCancelable(false);
        dialog.setCanceledOnTouchOutside(false);
        dialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
        dialog.getWindow().setLayout(
                ConstraintLayout.LayoutParams.MATCH_PARENT,
                ConstraintLayout.LayoutParams.WRAP_CONTENT
        );



        layoutBinding.bg.setOnClickListener(view -> dialog.dismiss());


        layoutBinding.cancel.setOnClickListener(view -> {
       ForgroundService.isDownload=false;
            Intent intent= new Intent(MainActivity.this, ForgroundService.class);
            stopService(intent);
            dialog.dismiss();
        });
        layoutBinding.progressBar.setMax(100);
}



    @Override
   public void itemClick(Item item) {
        Log.i("123321","item click");


        File file=new File(this.getApplicationInfo().dataDir +"/" +item.name+".pdf");
        if (file.exists()){
            startActivity(
                    new Intent(this, PdfActivity.class).putExtra(
                    "path",
                    file.getAbsolutePath()

            )
            );
        }
        else
        {
            if (ForgroundService.isDownload) {
                if (dialog != null) {
                    dialog.show();
                }
            }
            else
            {
                if (dialog!=null && !dialog.isShowing()) {
                    layoutBinding.progressBar.setIndeterminate(true);
                    layoutBinding.progressText.setText("--Mb/--Mb");
                    dialog.show();

                }
              new  Thread(() -> {

                  try {
                      URL myUrl =new URL(item.link);
                      URLConnection urlConnection = myUrl.openConnection();
                      urlConnection.connect();
                      int file_size = urlConnection.getContentLength();
                      String size = Utils.getBytesToMBString(file_size);
                      fileSize=Double.parseDouble(size);

                      Intent intent= new Intent(MainActivity.this,ForgroundService.class);
                   //   stopService(intent);


                      Intent startIntent = new Intent(MainActivity.this, ForgroundService.class);
                      startIntent.putExtra("name",item.getName()+".pdf");
                      startIntent.putExtra("size",fileSize);
                      startIntent.putExtra("link",item.getLink());
                      Log.i("123321", "startService: starting service");
                      ContextCompat.startForegroundService(MainActivity.this, startIntent);
                  }
                  catch ( IOException e) {
                      e.printStackTrace();
                  }
              }) {


            }.start();

            }
        }



    }

    @Override
    public void onNewIntent(Intent intent)
    {
        Log.i("123321", "onNewIntent: "+isComplete);
        super.onNewIntent(intent);
        if (!isComplete&&dialog!=null)dialog.show();
    }

    @Override
    protected void onStart() {
        isRunning=true;

        IntentFilter filter =new  IntentFilter();
        filter.addAction("timer_tracking");
        registerReceiver(mMessageReceiver,filter);
        super.onStart();
    }

    @Override
    protected void onDestroy() {
        isRunning=false;
        unregisterReceiver(mMessageReceiver);
        super.onDestroy();
    }

    @Override
    protected void onResume() {
        isRunning=true;
        super.onResume();
    }

    @Override
    protected void onPause() {
        isRunning=false;
        super.onPause();
    }
}

File: app/src/main/java/com/masum/pdfdownloader/MainActivity.kt
package com.masum.pdfdownloader

import android.app.Dialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.util.Log
import android.view.Window
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.LinearLayoutManager
import com.coolerfall.download.*
import com.masum.pdfdownloader.databinding.ActivityMainBinding
import com.masum.pdfdownloader.databinding.DownloadDialogBinding
import com.masum.pdfdownloader.others.Utils
import com.masum.pdfdownloader.services.ForegroundService
import java.io.File
import java.io.IOException
import java.net.URL
import java.net.URLConnection


class MainActivity : AppCompatActivity(),ItemClick{
    lateinit var binding: ActivityMainBinding
lateinit var list:ArrayList<Item>
var isRunning=false
    lateinit var layout:DownloadDialogBinding
    lateinit var dialog:Dialog
    lateinit var         mMessageReceiver: BroadcastReceiver
    var fileSize=0.0
    var isComplete=true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding= ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        layout= DownloadDialogBinding.inflate(layoutInflater)
        showDialog()

        list= ArrayList()
        list.add(Item("Test Name", "https://jsoncompare.org/LearningContainer/SampleFiles/PDF/sample-pdf-with-images.pdf"))
        list.add(Item("File Sample 150kb","https://file-examples-com.github.io/uploads/2017/10/file-sample_150kB.pdf"))
        list.add(Item("Rahe Belayet","https://www.pixelapps.info/my_important_files/pdf_files/rahe_belayat.pdf"))

        binding.recyclerview.layoutManager=LinearLayoutManager(this)
        binding.recyclerview.adapter=RecyclerItemAdapter(list,this)

     mMessageReceiver = object : BroadcastReceiver() {
       
            override fun onReceive(context: Context?, intent: Intent) {
                val current = intent.getDoubleExtra("timer", 0.0)
                val progress:Double=(current/(fileSize))*100
                if (dialog!=null&&layout!=null){
                    layout.progressBar.isIndeterminate=false
                    layout.progressText.text="$current Mb/${fileSize}Mb"
                    layout.progressBar.progress = progress.toInt()
                }


                val error:Boolean=intent.getBooleanExtra("error",false)
                if (error&&dialog!=null){
                    dialog.dismiss()
                    Toast.makeText(this@MainActivity,"Fail to download",Toast.LENGTH_SHORT).show()
                }
                val complete:Boolean=intent.getBooleanExtra("complete",false)
                if (complete){
                  if (binding.recyclerview.adapter!=null)  binding.recyclerview.adapter!!.notifyDataSetChanged()
                    isComplete=true
                    if (dialog!=null &&dialog.isShowing)
                        dialog.dismiss()
                    Log.i("123321", "onReceive: File download complete")

                    val file=File(intent.getStringExtra("path"))
                    if (file.exists())
                    {
                       if (isRunning&&dialog.isShowing){
                           startActivity(
                               Intent(this@MainActivity, MainActivity2::class.java).putExtra(
                                   "path",
                                   file.absolutePath
                                   //  holder.main.context.applicationInfo.dataDir.toString() + "/book.pdf"
                               )
                           )
                       }
                        else Toast.makeText(this@MainActivity,"Download Complete",Toast.LENGTH_SHORT).show()
                    }
                }
                else{
                    isComplete=false
                }






            }
        }


    }

    private fun showDialog() {
       dialog = Dialog(this)
        dialog?.requestWindowFeature(Window.FEATURE_NO_TITLE)

        dialog?.setContentView(layout.root)
        dialog?.setCancelable(false)
        dialog?.setCanceledOnTouchOutside(true)
        dialog?.window!!.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
        dialog.window!!.setLayout(
            ConstraintLayout.LayoutParams.MATCH_PARENT,
            ConstraintLayout.LayoutParams.WRAP_CONTENT
        )



    layout.bg.setOnClickListener {
        dialog.dismiss()
    }

        layout.cancel.setOnClickListener {
            ForegroundService.isDownload=false
          val intent=  Intent(this, ForegroundService::class.java)
            stopService(intent)
            dialog.dismiss()

        }
        layout.progressBar.max=100


    }

    override fun onStart() {
        isRunning=true
        val filter = IntentFilter()
        filter.addAction("timer_tracking")
        registerReceiver(mMessageReceiver,filter)
        super.onStart()
    }

    override fun onDestroy() {
        isRunning=false
        unregisterReceiver(mMessageReceiver)
        super.onDestroy()
    }

    override fun onResume() {
        isRunning=true
        super.onResume()
    }

    override fun onPause() {
       isRunning=false
        super.onPause()
    }

    override fun itemClick(item: Item) {


        val file=File(this.applicationInfo.dataDir.toString()+"/" +item.name+".pdf")
        if (file.exists()){
            startActivity(
                Intent(this, MainActivity2::class.java).putExtra(
                    "path",
                    file.absolutePath

                )
            )
        }
        else
        {
            Log.i("123321", "itemClick: ${ForegroundService.isDownload}")
            if (ForegroundService.isDownload) {
                if (dialog != null) {
                    dialog.show()
                }
            }
                else
                {
                    if (dialog!=null && !dialog.isShowing) {
                        layout.progressBar.isIndeterminate=true
                        layout.progressText.text="--Mb/--Mb"
                        dialog.show()
                    }
                    Thread {
                        try {
                            val myUrl = URL(item.link)
                            val urlConnection: URLConnection = myUrl.openConnection()
                            urlConnection.connect()
                            val file_size: Int = urlConnection.getContentLength()
                            val size = Utils.getBytesToMBString(file_size.toLong())
                            fileSize=size.toDouble()
                            val intent=  Intent(this, ForegroundService::class.java)
                            stopService(intent)


                            ForegroundService.startService(this@MainActivity, "${ item.name }.pdf",size.toDouble(),item.link)



                        } catch (e: IOException) {
                            e.printStackTrace()
                        }
                    }.start()

                }
            }



    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        Log.i("123321", "onNewIntent: ")
        if (!isComplete&&dialog!=null)dialog.show()
    }



}